File: 3962-maximum-subarray-sum-after-at-most-k-swaps/3962-maximum-subarray-sum-after-at-most-k-swaps.dart
class Solution {
  int maxSum(List<int> nums, int k) {
    int n = nums.length;

    int ans = -(1 << 63);

    if (k == 0 || n == 1) {
      int sm = 0;

      for (int num in nums) {
        sm += num;
        if (sm > ans) ans = sm;

        if (sm < 0) {
          sm = 0;
        }
      }

      return ans;
    }

    int cnt = 0;
    int cur = 0;
    List<int> pref = List.filled(n + 1, 0);

    for (int i = 0; i < n; i++) {
      if (nums[i] >= 0) {
        cur += nums[i];
      } else {
        cnt++;
      }

      pref[i + 1] = pref[i] + nums[i];

      if (nums[i] > ans) {
        ans = nums[i];
      }
    }

    if (cnt <= k) {
      return cur;
    }

    List<List<int>> dp =
        List.generate(n, (_) => List.filled(n, 0));

    for (int i = 0; i < n; i++) {
      List<int> pq = [];
      int sm = 0;

      for (int j = i; j < n; j++) {
        if (nums[j] >= 0) {
          dp[i][j] = sm;
          continue;
        }

        if (pq.length < k) {
          pq.add(nums[j]);
          pq.sort((a, b) => b.compareTo(a));
          sm += nums[j];
        } else if (pq[0] > nums[j]) {
          sm -= pq[0];
          pq[0] = nums[j];
          pq.sort((a, b) => b.compareTo(a));
          sm += nums[j];
        }
      }
    }

    for (int i = 0; i < n; i++) {
      List<int> pqMax = [];
      int sm = 0;

      for (int j = 0; j < i; j++) {
        if (nums[j] < 0) continue;

        if (pqMax.length < k) {
          pqMax.add(nums[j]);
          pqMax.sort();
          sm += nums[j];
        } else if (pqMax[0] < nums[j]) {
          sm -= pqMax[0];
          pqMax[0] = nums[j];
          pqMax.sort();
          sm += nums[j];
        }
      }

      for (int j = n - 1; j > i; j--) {
        int curSum = pref[j + 1] - pref[i];
        curSum -= dp[i][j];
        curSum += sm;

        if (curSum > ans) {
          ans = curSum;
        }

        if (nums[j] < 0) continue;

        if (pqMax.length < k) {
          pqMax.add(nums[j]);
          pqMax.sort();
          sm += nums[j];
        } else if (pqMax[0] < nums[j]) {
          sm -= pqMax[0];
          pqMax[0] = nums[j];
          pqMax.sort();
          sm += nums[j];
        }

        if (pqMax.isNotEmpty && sm > ans) {
          ans = sm;
        }
      }
    }

    return ans;
  }
}

File: 3962-maximum-subarray-sum-after-at-most-k-swaps/3962-maximum-subarray-sum-after-at-most-k-swaps.kt
class Solution {
    fun maxSum(nums: IntArray, k: Int): Long {

        val n = nums.size

        var ans = Int.MIN_VALUE.toLong()

        if (k == 0 || n == 1) {
            var sm = 0L

            for (num in nums) {
                sm += num.toLong()
                ans = maxOf(ans, sm)

                if (sm < 0)
                    sm = 0
            }

            return ans
        }

        var cnt = 0
        var cur = 0L
        val pref = LongArray(n + 1)

        for (i in 0 until n) {
            if (nums[i] >= 0)
                cur += nums[i].toLong()
            else
                cnt++

            pref[i + 1] = pref[i] + nums[i].toLong()
            ans = maxOf(ans, nums[i].toLong())
        }

        if (cnt <= k)
            return cur

        val dp = Array(n) { LongArray(n) }

        for (i in 0 until n) {
            val pq = java.util.PriorityQueue<Int>(reverseOrder())
            var sm = 0L

            for (j in i until n) {
                if (nums[j] >= 0) {
                    dp[i][j] = sm
                    continue
                }

                if (pq.size < k) {
                    pq.add(nums[j])
                    sm += nums[j].toLong()
                } else if (pq.peek() > nums[j]) {
                    sm -= pq.poll().toLong()
                    pq.add(nums[j])
                    sm += nums[j].toLong()
                }
            }
        }

        for (i in 0 until n) {
            val pqMax = java.util.PriorityQueue<Int>()
            var sm = 0L

            for (j in 0 until i) {
                if (nums[j] < 0)
                    continue

                if (pqMax.size < k) {
                    pqMax.add(nums[j])
                    sm += nums[j].toLong()
                } else if (pqMax.peek() < nums[j]) {
                    sm -= pqMax.poll().toLong()
                    pqMax.add(nums[j])
                    sm += nums[j].toLong()
                }
            }

            for (j in n - 1 downTo i + 1) {
                var curSum = pref[j + 1] - pref[i]
                curSum -= dp[i][j]
                curSum += sm

                ans = maxOf(ans, curSum)

                if (nums[j] < 0)
                    continue

                if (pqMax.size < k) {
                    pqMax.add(nums[j])
                    sm += nums[j].toLong()
                } else if (pqMax.peek() < nums[j]) {
                    sm -= pqMax.poll().toLong()
                    pqMax.add(nums[j])
                    sm += nums[j].toLong()
                }

                if (pqMax.isNotEmpty())
                    ans = maxOf(ans, sm)
            }
        }

        return ans
    }
}